File: packages/react-native-terminal/android/src/main/java/com/margelo/nitro/fressh/HybridTerminal.kt
package com.margelo.nitro.fressh

import android.view.Choreographer
import android.view.Surface
import android.view.SurfaceHolder
import android.view.SurfaceView
import android.view.View
import androidx.annotation.Keep
import com.facebook.proguard.annotations.DoNotStrip
import com.facebook.react.uimanager.ThemedReactContext

/**
 * Native terminal view (Nitro HybridView, the render plane §10). Owns a
 * [SurfaceView]; on surface creation it hands the `ANativeWindow` + the bound
 * `shellId` to the Rust render C-ABI (via JNI) and drives a Choreographer (vsync)
 * loop that draws the shell's durable `Term` from fressh-core's registry. The
 * byte stream never reaches JS. See docs §5/§10.
 */
@Keep
@DoNotStrip
class HybridTerminal(
  val context: ThemedReactContext,
) : HybridTerminalSpec() {
  // Plain SurfaceView: the buffer auto-tracks the view bounds. Keeping the grid in
  // lockstep with the on-screen size when the keyboard opens/closes is handled on the
  // Rust side, which polls the *settled* surface size from the draw loop (eglQuerySurface
  // lags the new geometry by a frame, so a one-shot read in surfaceChanged is unreliable
  // — esp. on GROW). The JS side (terminal flex:1 + the toolbar's keyboard-height-driven
  // marginBottom) is what actually resizes this view in both directions. See
  // docs/projects/complete/renderer-mismatched-selection-cutoff-scrollback.md.
  private val surfaceView = SurfaceView(context)

  /** Opaque pointer to the native render handle (0 = none). */
  private var nativeHandle: Long = 0L
  private var frameCallback: Choreographer.FrameCallback? = null

  override val view: View = surfaceView

  // Prop: bundled monospace font file path (no fontconfig on mobile, §6).
  override var fontPath: String = ""

  // Prop: render config as a JSON blob (physical px), assembled by the JS
  // <Terminal config={...}> wrapper. Carries color scheme, padding, cursor style,
  // bold-is-bright, font size. Live changes reflow the shell — a bonus over
  // desktop alacritty's restart-to-apply. We just forward the string; the renderer
  // parses + diffs it.
  override var configJson: String? = null
    set(value) {
      field = value
      if (nativeHandle != 0L) {
        nativeSetConfig(nativeHandle, value)
      }
    }

  // Prop: the durable shell to render. Rebinds the native view on change so the
  // same surface can follow a (re)started shell without a remount.
  override var shellId: String? = null
    set(value) {
      field = value
      if (nativeHandle != 0L) {
        nativeSetShell(nativeHandle, value)
      }
    }

  init {
    surfaceView.holder.addCallback(
      object : SurfaceHolder.Callback {
        override fun surfaceCreated(holder: SurfaceHolder) {
          nativeHandle = nativeAttach(holder.surface, resolveFontPath(), configJson, shellId)
          if (nativeHandle != 0L) startRenderLoop()
        }

        override fun surfaceChanged(holder: SurfaceHolder, format: Int, width: Int, height: Int) {
          if (nativeHandle != 0L) nativeResize(nativeHandle)
        }

        override fun surfaceDestroyed(holder: SurfaceHolder) {
          stopRenderLoop()
          if (nativeHandle != 0L) {
            nativeDestroy(nativeHandle)
            nativeHandle = 0L
          }
        }
      },
    )
  }

  /**
   * The native renderer needs a font *file* path. If RN didn't supply one, fall
   * back to the bundled DejaVu Sans Mono asset, extracted to filesDir once.
   */
  private fun resolveFontPath(): String {
    if (fontPath.isNotEmpty()) return fontPath
    val out = java.io.File(context.filesDir, "DejaVuSansMono.ttf")
    if (!out.exists()) {
      context.assets.open("fonts/DejaVuSansMono.ttf").use { input ->
        out.outputStream().use { output -> input.copyTo(output) }
      }
    }
    return out.absolutePath
  }

  private fun startRenderLoop() {
    val callback =
      object : Choreographer.FrameCallback {
        override fun doFrame(frameTimeNanos: Long) {
          if (nativeHandle == 0L) return
          nativeDraw(nativeHandle)
          Choreographer.getInstance().postFrameCallback(this)
        }
      }
    frameCallback = callback
    Choreographer.getInstance().postFrameCallback(callback)
  }

  private fun stopRenderLoop() {
    frameCallback?.let { Choreographer.getInstance().removeFrameCallback(it) }
    frameCallback = null
  }

  // JNI bridge -> Rust render C-ABI (see android/src/main/cpp/cpp-adapter.cpp).
  private external fun nativeAttach(
    surface: Surface,
    fontPath: String,
    configJson: String?,
    shellId: String?,
  ): Long

  private external fun nativeSetShell(handle: Long, shellId: String?)

  private external fun nativeSetConfig(handle: Long, configJson: String?)

  private external fun nativeDraw(handle: Long)

  private external fun nativeResize(handle: Long)

  private external fun nativeSendInput(handle: Long, data: ByteArray)

  private external fun nativeDestroy(handle: Long)

  companion object {
    init {
      System.loadLibrary("ReactNativeTerminal")
    }
  }
}

File: packages/react-native-terminal/ios/HybridTerminal.swift
import NitroModules
import QuartzCore
import UIKit

/// CAMetalLayer-backed view that drives the Rust render C-ABI (ANGLE→Metal, §5) —
/// the iOS analogue of HybridTerminal.kt's `SurfaceView` + `Choreographer` loop.
///
/// On first layout it `fressh_terminal_attach`es its `CAMetalLayer` to ANGLE,
/// runs a `CADisplayLink` (vsync) loop → `fressh_terminal_draw` (which looks the
/// shell's durable `Term` up from fressh-core's registry by `shellId` and draws it
/// — the byte stream never reaches JS), and forwards shellId/config/resize/teardown
/// to the C-ABI. Changing `shellId` rebinds (instant reattach, full scrollback, §9).
final class TerminalMetalView: UIView {
  // The view's backing layer IS a CAMetalLayer — what ANGLE renders into via Metal.
  override class var layerClass: AnyClass { CAMetalLayer.self }
  private var metalLayer: CAMetalLayer { layer as! CAMetalLayer }

  /// Opaque Rust render handle (nil = not attached).
  private var handle: UnsafeMutableRawPointer?
  private var displayLink: CADisplayLink?

  /// Bundled monospace font file path (no fontconfig on mobile, §6/§8).
  var fontPath: String = ""

  /// Render config JSON (physical px), assembled by the JS `<Terminal config>` wrapper.
  var configJson: String? {
    didSet {
      guard let handle else { return }
      withOptCString(configJson) { fressh_terminal_set_config(handle, $0) }
    }
  }

  /// The durable shell to render. Rebinds the live view without a remount.
  var shellId: String? {
    didSet {
      guard let handle else { return }
      withOptCString(shellId) { fressh_terminal_set_shell(handle, $0) }
    }
  }

  override func didMoveToWindow() {
    super.didMoveToWindow()
    if window == nil { teardown() } else { attachIfNeeded() }
  }

  override func layoutSubviews() {
    super.layoutSubviews()
    syncDrawableSize()
    // Attach once we have real bounds; afterwards just keep the grid in lockstep
    // with the on-screen size (keyboard open/close), mirroring the Android resize poll.
    if handle != nil {
      fressh_terminal_resize(handle)
    } else {
      attachIfNeeded()
    }
  }

  /// Keep the Metal drawable in physical px (logical bounds × screen scale).
  private func syncDrawableSize() {
    let scale = window?.screen.scale ?? UIScreen.main.scale
    metalLayer.contentsScale = scale
    let size = CGSize(width: bounds.width * scale, height: bounds.height * scale)
    if size.width > 0, size.height > 0 { metalLayer.drawableSize = size }
  }

  private func attachIfNeeded() {
    guard handle == nil, window != nil, bounds.width > 0, bounds.height > 0 else { return }
    syncDrawableSize()
    let layerPtr = Unmanaged.passUnretained(metalLayer).toOpaque()
    handle = resolveFontPath().withCString { fp in
      withOptCString(configJson) { cfg in
        withOptCString(shellId) { sid in
          fressh_terminal_attach(layerPtr, fp, cfg, sid)
        }
      }
    }
    if handle != nil { startRenderLoop() }
  }

  private func startRenderLoop() {
    let link = CADisplayLink(target: self, selector: #selector(onFrame))
    link.add(to: .main, forMode: .common)
    displayLink = link
  }

  @objc private func onFrame() {
    guard let handle else { return }
    fressh_terminal_draw(handle)
  }

  /// Drop the renderer/EGL context. The shell's `Term` survives in the registry,
  /// so a later attach to the same `shellId` resumes instantly (§9).
  func teardown() {
    displayLink?.invalidate()
    displayLink = nil
    if let handle {
      fressh_terminal_destroy(handle)
      self.handle = nil
    }
  }

  /// The native renderer needs a font *file* path. Fall back to the bundled
  /// DejaVu Sans Mono shipped in the podspec resource bundle.
  private func resolveFontPath() -> String {
    if !fontPath.isEmpty { return fontPath }
    let owner = Bundle(for: TerminalMetalView.self)
    if let bundleURL = owner.url(forResource: "FresshTerminalFonts", withExtension: "bundle"),
      let resources = Bundle(url: bundleURL),
      let font = resources.path(forResource: "DejaVuSansMono", ofType: "ttf")
    {
      return font
    }
    return owner.path(forResource: "DejaVuSansMono", ofType: "ttf") ?? ""
  }

  deinit { teardown() }
}

/// Call `body` with a C string for `s`, or NULL when `s` is nil.
private func withOptCString<R>(_ s: String?, _ body: (UnsafePointer<CChar>?) -> R) -> R {
  if let s { return s.withCString { body($0) } }
  return body(nil)
}

/// iOS implementation of the `Terminal` Nitro HybridView (the render plane, §10).
/// Owns the `CAMetalLayer` view and forwards props to it.
class HybridTerminal: HybridTerminalSpec {
  private let metalView = TerminalMetalView()
  var view: UIView { metalView }

  var fontPath: String = "" {
    didSet { metalView.fontPath = fontPath }
  }
  var configJson: String? {
    didSet { metalView.configJson = configJson }
  }
  var shellId: String? {
    didSet { metalView.shellId = shellId }
  }

  func onDropView() {
    metalView.teardown()
  }
}
